File: 0844-backspace-string-compare/0844-backspace-string-compare.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        i, j = len(s)-1, len(t)-1
        back_s, back_t = 0, 0
        while True:
          while i>=0 and (s[i]=='#' or back_s):
            back_s += 1 if s[i]=='#' else -1
            i -= 1
          while j>=0 and (t[j]=='#' or back_t):
            back_t += 1 if t[j]=='#' else -1
            j -= 1
          if not (i>=0 and j>=0 and s[i]==t[j]):
            return i == -1 and j == -1
          i, j = i-1, j-1
        return True

File: 0844-backspace-string-compare/0844-backspace-string-compare.cpp
class Solution {
public:
    bool backspaceCompare(string s, string t) {
        int m { static_cast<int>(s.length())}, n { static_cast<int>(t.length())};
        int s_skip {}, t_skip {};
        int i { m-1 }, j { n-1 };
        while (i>=0 || j>=0) {
            while (i>=0) {
                if (s[i] == '#') s_skip++;
                else if (s_skip) s_skip--;
                else break;
                i--;
            }
            while (j>=0) {
                if (t[j] == '#') t_skip++;
                else if (t_skip) t_skip--;
                else break;
                j--;
            }
            if (i<0 != j<0) return false;
            if (i>=0 && s[i--]!=t[j--]) return false;
        }
        return true;
    }
};